--- IDocumentLink.h	2008-03-20 19:16:03.000000000 +0100
+++ epdfview-1.6-r1/src/IDocumentLink.h	2008-03-20 19:16:43.000000000 +0100
@@ -33,7 +33,7 @@
     class IDocumentLink
     {
         public:
-            IDocumentLink (gdouble x1, gdouble y1, gdouble x2, gdouble y1);
+            IDocumentLink (gdouble x1, gdouble y1, gdouble x2, gdouble y2);
             virtual ~IDocumentLink (void);
 
             virtual void activate (IDocument *document) = 0;
