--- src/SConstruct	2012-01-28 20:33:17.465511337 +0100
+++ src/SConstruct	2012-01-28 20:34:52.736320291 +0100
@@ -30,9 +30,8 @@
 f.close()
 
 env.CacheDir('/tmp/.zfs-fuse.scons')
-env['CC'] = 'gcc'
-env['LINKFLAGS'] = Split('-pipe -Wall')
-env['CCFLAGS'] = Split('-pipe -Wall -std=c99 -Wno-switch -Wno-unused -Wno-missing-braces -Wno-parentheses -Wno-uninitialized -fno-strict-aliasing -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DTEXT_DOMAIN=\\"zfs-fuse\\" ')
+env['LINKFLAGS'] = Split(os.environ.get('LDFLAGS', ''))
+env['CCFLAGS'] = Split(os.environ.get('CFLAGS', '') + ' -Wall -std=c99 -Wno-switch -Wno-unused -Wno-missing-braces -Wno-parentheses -Wno-uninitialized -fno-strict-aliasing -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DTEXT_DOMAIN=\\"zfs-fuse\\" ')
 
 if osname == "Linux":
   env.Append(CPPFLAGS = " -DLINUX_AIO")
