diff -urN genkernel-3.4.45-a/gen_initramfs.sh genkernel-3.4.45-b/gen_initramfs.sh
--- genkernel-3.4.45-a/gen_initramfs.sh	2012-11-06 06:32:57.000000000 +0000
+++ genkernel-3.4.45-b/gen_initramfs.sh	2013-01-22 21:34:31.229278294 +0000
@@ -585,11 +585,13 @@
 
 	cp -ax --parents ./lib/modules/${KV}/modules* ${TEMP}/initramfs-modules-${KV}-temp 2>/dev/null
 
-	mkdir -p "${TEMP}/initramfs-modules-${KV}-temp/etc/modules"
-	for group_modules in ${!MODULES_*}; do
-		group="$(echo $group_modules | cut -d_ -f2 | tr "[:upper:]" "[:lower:]")"
-		print_list ${!group_modules} > "${TEMP}/initramfs-modules-${KV}-temp/etc/modules/${group}"
-	done
+    find ${TEMP}/initramfs-modules-${KV}-temp -name "*.ko" -exec strip --strip-unneeded '{}' \;
+
+	#mkdir -p "${TEMP}/initramfs-modules-${KV}-temp/etc/modules"
+	#for group_modules in ${!MODULES_*}; do
+	#	group="$(echo $group_modules | cut -d_ -f2 | tr "[:upper:]" "[:lower:]")"
+	#	print_list ${!group_modules} > "${TEMP}/initramfs-modules-${KV}-temp/etc/modules/${group}"
+	#done
 	cd "${TEMP}/initramfs-modules-${KV}-temp/"
 	log_future_cpio_content
 	find . | cpio ${CPIO_ARGS} --append -F "${CPIO}" \
@@ -759,17 +761,17 @@
 	echo | cpio ${CPIO_ARGS} -F "${CPIO}" 2>/dev/null \
 		|| gen_die "Could not create empty cpio at ${CPIO}"
 
-	append_data 'base_layout'
-	append_data 'auxilary' "${BUSYBOX}"
-	append_data 'busybox' "${BUSYBOX}"
-	isTrue "${CMD_E2FSPROGS}" && append_data 'e2fsprogs'
-	append_data 'lvm' "${LVM}"
-	append_data 'dmraid' "${DMRAID}"
-	append_data 'iscsi' "${ISCSI}"
-	append_data 'mdadm' "${MDADM}"
-	append_data 'luks' "${LUKS}"
-	append_data 'multipath' "${MULTIPATH}"
-	append_data 'gpg' "${GPG}"
+	#append_data 'base_layout'
+	#append_data 'auxilary' "${BUSYBOX}"
+	#append_data 'busybox' "${BUSYBOX}"
+	#isTrue "${CMD_E2FSPROGS}" && append_data 'e2fsprogs'
+	#append_data 'lvm' "${LVM}"
+	#append_data 'dmraid' "${DMRAID}"
+	#append_data 'iscsi' "${ISCSI}"
+	#append_data 'mdadm' "${MDADM}"
+	#append_data 'luks' "${LUKS}"
+	#append_data 'multipath' "${MULTIPATH}"
+	#append_data 'gpg' "${GPG}"
 
 	if [ "${RAMDISKMODULES}" = '1' ]
 	then
@@ -778,15 +780,15 @@
 		print_info 1 "initramfs: Not copying modules..."
 	fi
 
-	append_data 'zfs' "${ZFS}"
+	#append_data 'zfs' "${ZFS}"
 
-	append_data 'blkid' "${DISKLABEL}"
+	#append_data 'blkid' "${DISKLABEL}"
 
-	append_data 'unionfs_fuse' "${UNIONFS}"
+	#append_data 'unionfs_fuse' "${UNIONFS}"
 
-	append_data 'splash' "${SPLASH}"
+	#append_data 'splash' "${SPLASH}"
 
-	append_data 'modprobed'
+	#append_data 'modprobed'
 
 	if isTrue "${FIRMWARE}" && [ -n "${FIRMWARE_DIR}" ]
 	then
@@ -801,6 +803,21 @@
 
 	if isTrue "${INTEGRATED_INITRAMFS}"
 	then
+        # Compress modules with gzip to save memory when initramfs is uncompressed
+		print_info 1 "        >> Gzipping kernel modules in embedded initramfs ..."
+		INITTMP="${TMPDIR}/embedded-initramfs"
+		BAKARXZ="/etc/kernels/initramfs-${KV}${EXV}.cpio.xz"
+		mkdir -p ${INITTMP}
+		( cd ${INITTMP} ; cat ${CPIO} | cpio -id --quiet )
+		for curdir in ${INITTMP}/lib*/modules/
+		do
+			find ${curdir} -iname "*ko" -exec gzip -9 {} \;
+		done
+		( cd ${INITTMP} ; find . | cpio -o -H newc --quiet >| ${CPIO} )
+		print_info 1 "        >> Embedded initramfs archive created in ${CPIO}"
+		print_info 1 "        >> Copying embedded initramfs to ${BAKARXZ}"
+		cat ${CPIO} | xz >| "${BAKARXZ}"
+		rm -rf ${INITTMP}
 		# Explicitly do not compress if we are integrating into the kernel.
 		# The kernel will do a better job of it than us.
 		mv ${TMPDIR}/initramfs-${KV} ${TMPDIR}/initramfs-${KV}.cpio
