--- src/main.c.orig	2010-12-12 04:41:48.000000000 +0800
+++ src/main.c	2010-12-16 12:09:51.482000212 +0800
@@ -185,8 +185,6 @@
     if (!mm_options_debug ())
         logging_setup ();
 
-    g_message ("ModemManager (version " MM_DIST_VERSION ") starting...");
-
     bus = dbus_g_bus_get (DBUS_BUS_SYSTEM, &err);
     if (!bus) {
         g_warning ("Could not get the system bus. Make sure "
--- src/mm-manager.c.orig	2010-12-12 04:41:48.000000000 +0800
+++ src/mm-manager.c	2010-12-16 12:10:37.499999988 +0800
@@ -112,7 +112,6 @@
     plugin = (*plugin_create_func) ();
     if (plugin) {
         g_object_weak_ref (G_OBJECT (plugin), (GWeakNotify) g_module_close, module);
-        g_message ("Loaded plugin %s", mm_plugin_get_name (plugin));
     } else
         g_warning ("Could not load plugin %s: initialization failed", path);
 
@@ -781,7 +780,8 @@
         if (   strcmp (name, "console")
             && strcmp (name, "ptmx")
             && strcmp (name, "lo")
-            && strcmp (name, "tty")
+            && strncmp (name, "tty", 3)
+            && strncmp (name, "pty", 3)
             && !strstr (name, "virbr"))
             g_debug ("(%s/%s): could not get port's parent device", subsys, name);
 
