diff -urN ModemManager-0.4_p20101211/src/main.c ModemManager-0.4_p20101211-patch/src/main.c
--- ModemManager-0.4_p20101211/src/main.c	2010-12-11 20:41:48.000000000 +0000
+++ ModemManager-0.4_p20101211-patch/src/main.c	2010-12-26 12:55:10.000000000 +0000
@@ -185,8 +185,6 @@
     if (!mm_options_debug ())
         logging_setup ();
 
-    g_message ("ModemManager (version " MM_DIST_VERSION ") starting...");
-
     bus = dbus_g_bus_get (DBUS_BUS_SYSTEM, &err);
     if (!bus) {
         g_warning ("Could not get the system bus. Make sure "
diff -urN ModemManager-0.4_p20101211/src/mm-manager.c ModemManager-0.4_p20101211-patch/src/mm-manager.c
--- ModemManager-0.4_p20101211/src/mm-manager.c	2010-12-11 20:41:48.000000000 +0000
+++ ModemManager-0.4_p20101211-patch/src/mm-manager.c	2010-12-26 12:57:09.000000000 +0000
@@ -112,7 +112,6 @@
     plugin = (*plugin_create_func) ();
     if (plugin) {
         g_object_weak_ref (G_OBJECT (plugin), (GWeakNotify) g_module_close, module);
-        g_message ("Loaded plugin %s", mm_plugin_get_name (plugin));
     } else
         g_warning ("Could not load plugin %s: initialization failed", path);
 
@@ -778,19 +777,11 @@
         /* Warn about it, but filter out some common ports that we know don't have
          * anything to do with mobile broadband.
          */
-        if (   strcmp (name, "console")
-            && strcmp (name, "ptmx")
-            && strcmp (name, "lo")
-            && strcmp (name, "tty")
-            && !strstr (name, "virbr"))
-            g_debug ("(%s/%s): could not get port's parent device", subsys, name);
-
         goto out;
     }
 
     /* Is the device blacklisted? */
     if (g_udev_device_get_property_as_boolean (physdev, "ID_MM_DEVICE_IGNORE")) {
-        g_debug ("(%s/%s): port's parent device is blacklisted", subsys, name);
         goto out;
     }
 
@@ -799,13 +790,11 @@
     if (   physdev_subsys
         && !strcmp (physdev_subsys, "platform")
         && !g_udev_device_get_property_as_boolean (physdev, "ID_MM_PLATFORM_DRIVER_PROBE")) {
-        g_debug ("(%s/%s): port's parent platform driver is not whitelisted", subsys, name);
         goto out;
     }
 
     physdev_path = g_udev_device_get_sysfs_path (physdev);
     if (!physdev_path) {
-        g_debug ("(%s/%s): could not get port's parent device sysfs path", subsys, name);
         goto out;
     }
 
