diff -urN ModemManager-0.4_p20101211-a/src/mm-at-serial-port.h ModemManager-0.4_p20101211-b/src/mm-at-serial-port.h
--- ModemManager-0.4_p20101211-a/src/mm-at-serial-port.h	2010-12-11 20:41:48.000000000 +0000
+++ ModemManager-0.4_p20101211-b/src/mm-at-serial-port.h	2012-10-10 20:54:03.693303644 +0100
@@ -18,7 +18,7 @@
 #define MM_AT_SERIAL_PORT_H
 
 #include <glib.h>
-#include <glib/gtypes.h>
+#include <glib.h>
 #include <glib-object.h>
 
 #include "mm-serial-port.h"
diff -urN ModemManager-0.4_p20101211-a/src/mm-manager.h ModemManager-0.4_p20101211-b/src/mm-manager.h
--- ModemManager-0.4_p20101211-a/src/mm-manager.h	2010-12-11 20:41:48.000000000 +0000
+++ ModemManager-0.4_p20101211-b/src/mm-manager.h	2012-10-10 20:54:03.695304909 +0100
@@ -17,7 +17,7 @@
 #ifndef MM_MANAGER_H
 #define MM_MANAGER_H
 
-#include <glib/gtypes.h>
+#include <glib.h>
 #include <glib-object.h>
 #include <dbus/dbus-glib.h>
 #include "mm-modem.h"
diff -urN ModemManager-0.4_p20101211-a/src/mm-modem-base.h ModemManager-0.4_p20101211-b/src/mm-modem-base.h
--- ModemManager-0.4_p20101211-a/src/mm-modem-base.h	2010-12-11 20:41:48.000000000 +0000
+++ ModemManager-0.4_p20101211-b/src/mm-modem-base.h	2012-10-10 20:54:03.696304824 +0100
@@ -18,7 +18,7 @@
 #define MM_MODEM_BASE_H
 
 #include <glib.h>
-#include <glib/gtypes.h>
+#include <glib.h>
 #include <glib-object.h>
 
 #include "mm-port.h"
diff -urN ModemManager-0.4_p20101211-a/src/mm-plugin-base.h ModemManager-0.4_p20101211-b/src/mm-plugin-base.h
--- ModemManager-0.4_p20101211-a/src/mm-plugin-base.h	2010-12-11 20:41:48.000000000 +0000
+++ ModemManager-0.4_p20101211-b/src/mm-plugin-base.h	2012-10-10 20:54:03.699304899 +0100
@@ -17,7 +17,7 @@
 #define MM_PLUGIN_BASE_H
 
 #include <glib.h>
-#include <glib/gtypes.h>
+#include <glib.h>
 #include <glib-object.h>
 
 #define G_UDEV_API_IS_SUBJECT_TO_CHANGE
diff -urN ModemManager-0.4_p20101211-a/src/mm-port.h ModemManager-0.4_p20101211-b/src/mm-port.h
--- ModemManager-0.4_p20101211-a/src/mm-port.h	2010-12-11 20:41:48.000000000 +0000
+++ ModemManager-0.4_p20101211-b/src/mm-port.h	2012-10-10 20:54:12.676427708 +0100
@@ -17,7 +17,6 @@
 #define MM_PORT_H
 
 #include <glib.h>
-#include <glib/gtypes.h>
 #include <glib-object.h>
 
 typedef enum {
diff -urN ModemManager-0.4_p20101211-a/src/mm-qcdm-serial-port.h ModemManager-0.4_p20101211-b/src/mm-qcdm-serial-port.h
--- ModemManager-0.4_p20101211-a/src/mm-qcdm-serial-port.h	2010-12-11 20:41:48.000000000 +0000
+++ ModemManager-0.4_p20101211-b/src/mm-qcdm-serial-port.h	2012-10-10 20:54:03.700304844 +0100
@@ -18,7 +18,7 @@
 #define MM_QCDM_SERIAL_PORT_H
 
 #include <glib.h>
-#include <glib/gtypes.h>
+#include <glib.h>
 #include <glib-object.h>
 
 #include "mm-serial-port.h"
diff -urN ModemManager-0.4_p20101211-a/src/mm-serial-port.h ModemManager-0.4_p20101211-b/src/mm-serial-port.h
--- ModemManager-0.4_p20101211-a/src/mm-serial-port.h	2010-12-11 20:41:48.000000000 +0000
+++ ModemManager-0.4_p20101211-b/src/mm-serial-port.h	2012-10-10 20:54:03.700304844 +0100
@@ -18,7 +18,7 @@
 #define MM_SERIAL_PORT_H
 
 #include <glib.h>
-#include <glib/gtypes.h>
+#include <glib.h>
 #include <glib-object.h>
 
 #include "mm-port.h"
