diff -urN ModemManager-0.6.0.0-a/src/main.c ModemManager-0.6.0.0-b/src/main.c
--- ModemManager-0.6.0.0-a/src/main.c	2012-08-29 05:08:16.000000000 +0100
+++ ModemManager-0.6.0.0-b/src/main.c	2012-11-11 13:43:23.029897000 +0000
@@ -129,7 +129,7 @@
 
     GOptionEntry entries[] = {
 		{ "debug", 0, 0, G_OPTION_ARG_NONE, &debug, "Output to console rather than syslog", NULL },
-		{ "log-level", 0, 0, G_OPTION_ARG_STRING, &log_level, "Log level: one of [ERR, WARN, INFO, DEBUG]", "INFO" },
+		{ "log-level", 0, 0, G_OPTION_ARG_STRING, &log_level, "Log level: one of [ERR, WARN, INFO, DEBUG]", "WARN" },
 		{ "log-file", 0, 0, G_OPTION_ARG_STRING, &log_file, "Path to log file", NULL },
 		{ "timestamps", 0, 0, G_OPTION_ARG_NONE, &show_ts, "Show timestamps in log output", NULL },
 		{ "relative-timestamps", 0, 0, G_OPTION_ARG_NONE, &rel_ts, "Use relative timestamps (from MM start)", NULL },
@@ -164,7 +164,7 @@
 
     setup_signals ();
 
-    mm_info ("ModemManager (version " MM_DIST_VERSION ") starting...");
+    /*mm_info ("ModemManager (version " MM_DIST_VERSION ") starting...");*/
 
     bus = dbus_g_bus_get (DBUS_BUS_SYSTEM, &err);
     if (!bus) {
diff -urN ModemManager-0.6.0.0-a/src/mm-manager.c ModemManager-0.6.0.0-b/src/mm-manager.c
--- ModemManager-0.6.0.0-a/src/mm-manager.c	2012-08-29 15:58:11.000000000 +0100
+++ ModemManager-0.6.0.0-b/src/mm-manager.c	2012-11-11 13:43:00.869897000 +0000
@@ -155,7 +155,7 @@
 static void
 found_plugin (MMPlugin *plugin)
 {
-    mm_info ("Loaded plugin '%s'", mm_plugin_get_name (plugin));
+    /*mm_info ("Loaded plugin '%s'", mm_plugin_get_name (plugin));*/
 }
 
 static void
@@ -207,7 +207,7 @@
      * used while checking support */
     g_slist_foreach (priv->plugins, (GFunc)found_plugin, NULL);
 
-    mm_info ("Successfully loaded %u plugins", g_slist_length (priv->plugins));
+    /*mm_info ("Successfully loaded %u plugins", g_slist_length (priv->plugins));*/
 
     g_dir_close (dir);
 }
