diff -urN squashfs-4.2-a/squashfs-tools/compressor.c squashfs-4.2-b/squashfs-tools/compressor.c
--- squashfs-4.2-a/squashfs-tools/compressor.c	2011-02-11 15:49:24.000000000 +0000
+++ squashfs-4.2-b/squashfs-tools/compressor.c	2011-04-06 07:28:13.000000000 +0100
@@ -27,7 +27,7 @@
 
 #ifndef GZIP_SUPPORT
 static struct compressor gzip_comp_ops =  {
-	NULL, NULL, NULL, NULL, NULL, NULL, ZLIB_COMPRESSION, "gzip", 0
+	NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, ZLIB_COMPRESSION, "gzip", 0
 };
 #else
 extern struct compressor gzip_comp_ops;
