diff -urN parted-2.2-orig/libparted/arch/linux.c parted-2.2-patch/libparted/arch/linux.c
--- parted-2.2-orig/libparted/arch/linux.c	2010-02-22 08:04:53.000000000 +0000
+++ parted-2.2-patch/libparted/arch/linux.c	2010-03-22 18:51:02.000000000 +0000
@@ -2521,15 +2521,13 @@
 _kernel_reread_part_table (PedDevice* dev)
 {
         LinuxSpecific*  arch_specific = LINUX_SPECIFIC (dev);
-        int             retry_count = 9;
+        int             retry_count = 70;
 
         sync();
         while (ioctl (arch_specific->fd, BLKRRPART)) {
                 retry_count--;
                 sync();
-                if (retry_count == 3)
-                        sleep(1); /* Pause to allow system to settle */
-
+		usleep(100000); /* sleep(0.1) to allow system to settle */
                 if (!retry_count) {
                         ped_exception_throw (
                                 PED_EXCEPTION_WARNING,
